calico/felix/test/test_endpoint.py

```diff
@@ -32,21 +33,153 @@
+        self.m_config.IFACE_PREFIX = "tap"
+        self.m_iptables_updater = Mock(spec=IptablesUpdater)
+        self.m_dispatch_chains = Mock(spec=DispatchChains)
+        self.m_rules_mgr = Mock(spec=RulesManager)
+
+    def get_local_endpoint(self, combined_id, ip_type):
+        local_endpoint = endpoint.LocalEndpoint(self.m_config,
+                                                combined_id,
+                                                ip_type,
+                                                self.m_iptables_updater,
+                                                self.m_dispatch_chains,
+                                                self.m_rules_mgr)
+
+        # For purposes of our testing, we force things to happen in line.
+        local_endpoint.greenlet = gevent.getcurrent()
+        return local_endpoint
+
+    def test_on_endpoint_update_v4(self):
+        combined_id = EndpointId("host_id", "orchestrator_id",
+                                 "workload_id", "endpoint_id")
+        ip_type = futils.IPV4
+        retcode = futils.CommandOutput("", "")
+        local_ep = self.get_local_endpoint(combined_id, ip_type)
+
+        # Call with no data.
+        result = local_ep.on_endpoint_update(None, async=False)
+
+        ips = ["1.2.3.4"]
+        iface = "tapabcdef"
+        data = { 'endpoint': "endpoint_id",
+                 'mac': stub_utils.get_mac(),
+                 'name': iface,
+                 'ipv4_nets': ips,
+                 'profile_ids': []}
+
+        with mock.patch('calico.felix.devices.set_routes'):
+            with mock.patch('calico.felix.devices.configure_interface_ipv4'):
+                result = local_ep.on_endpoint_update(data, async=False)
+                self.assertEqual(local_ep._mac, data['mac'])
+                devices.configure_interface_ipv4.assert_called_once_with(iface)
+                devices.set_routes.assert_called_once_with(ip_type,
+                                                           set(ips),
+                                                           iface,
+                                                           data['mac'],
+                                                           reset_arp=True)
+
+        with mock.patch('calico.felix.devices.set_routes'):
+            with mock.patch('calico.felix.devices.configure_interface_ipv4'):
+                result = local_ep.on_endpoint_update(data, async=False)
+                self.assertEqual(local_ep._mac, data['mac'])
+                devices.configure_interface_ipv4.assert_called_once_with(iface)
+                devices.set_routes.assert_called_once_with(ip_type,
+                                                           set(ips),
+                                                           iface,
+                                                           data['mac'],
+                                                           reset_arp=False)
+
+        data['mac'] = stub_utils.get_mac()
```

Comment:
These blocks are pretty scary looking, and it's not clear what the test is doing unless you know exactly what the code   it's testing does.  Please add comments (short ones are fine) to say what each step is testing.